### PR TITLE
Add basic security functions to druidapi (follow up to #14009)

### PR DESCRIPTION
**Pull request by abhishekrb19**
Pull in @paul-rogers's changes from https://github.com/apache/druid/pull/14009. This PR addresses the review feedback from the PR and applies some more additional changes, including:
- Add ```bash and remove `$>` from codeblocks so they can be copy-pasted as-is or run from an IDE.
-  Fix a typo

This PR has:

- [x] been self-reviewed.
- [x] added documentation for new or modified features or behaviors.
(https://github.com/apache/druid/blob/master/dev/license.md)

### CODE REVIEW

File: docs/operations/security-overview.md [hunk 1]
```diff
@@ -176,70 +169,83 @@ An example configuration:
...
+   replacing USERNAME with the *new* username you are trying to create. For example:
+   ```bash
    curl -u admin:password1 -XPOST https://my-coordinator-ip:8281/druid-ext/basic-security/authentication/db/MyBasicMetadataAuthenticator/users/myname
-  ```
-  >  If you have TLS enabled, be sure to adjust the curl command accordingly. For example, if your Druid servers use self-signed certificates, you may choose to include the `insecure` curl option to forgo certificate checking for the curl command. 
+   ```
+   >  If you have TLS enabled, be sure to adjust the curl command accordingly. For example, if your Druid servers use self-signed certificates,
+   you may choose to include the `insecure` curl option to forgo certificate checking for the curl command.
+
 2. Add a credential for the user by issuing a POST to `druid-ext/basic-security/authentication/db/MyBasicMetadataAuthenticator/users/<USERNAME>/credentials`. For example:
```
Comment (vtlim) [inline docs/operations/security-overview.md:208]:
```suggestion
2. Add a credential for the user by issuing a POST request to `druid-ext/basic-security/authentication/db/MyBasicMetadataAuthenticator/users/<USERNAME>/credentials`. For example:
```

File: docs/operations/security-overview.md [hunk 2]
```diff
@@ -176,70 +169,83 @@ An example configuration:
...
+5. Assign roles to users by issuing a POST request to `druid-ext/basic-security/authorization/db/MyBasicMetadataAuthorizer/users/<USERNAME>/roles/<ROLENAME>`. For example:
+   ```bash
+   curl -u admin:password1 -XPOST https://my-coordinator-ip:8281/druid-ext/basic-security/authorization/db/MyBasicMetadataAuthorizer/users/myname/roles/myrole | jq
+   ```
+
+6. Finally, attach permissions to the roles to control how they can interact with Druid at `druid-ext/basic-security/authorization/db/MyBasicMetadataAuthorizer/roles/<ROLENAME>/permissions`. For example:
+   ```bash
+   curl -u admin:password1 -H'Content-Type: application/json' -XPOST --data-binary @perms.json https://my-coordinator-ip:8281/druid-ext/basic-security/authorization/db/MyBasicMetadataAuthorizer/roles/myrole/permissions
+   ```
+   payload of `perms.json` should be in the form:
```
Comment (vtlim) [inline docs/operations/security-overview.md:229]:
```suggestion
   The payload of `perms.json` should be in the following form:
```

File: examples/quickstart/jupyter-notebooks/druidapi/druidapi/basic_auth.py [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+    ac.add_role('myRole')
+
+    # Assign the role to the user
+    ac.assign_role_to_user('myRole', 'bob')
+
+    # Give the role some permissions
+    ac.grant_permissions('myRole', [[consts.DATASOURCE_RESOURCE, 'foo', consts.READ_ACTION]])
+    ```
+
+    Then use the various other methods to list users, roles and permissions to verify the
```
Comment (vtlim) [inline examples/quickstart/jupyter-notebooks/druidapi/druidapi/basic_auth.py:73]:
```suggestion
    Then use the various other methods to list users, roles, and permissions to verify the
```

File: docs/operations/security-overview.md [hunk 4]
```diff
@@ -157,10 +148,12 @@ An example configuration:
    # Default password for internal 'druid_system' user, should be changed for production.
    druid.auth.authenticator.MyBasicMetadataAuthenticator.initialInternalClientPassword=password2
 
-   # Uses the metadata store for storing users, you can use authentication API to create new users and grant permissions
+   # Uses the metadata store for storing users.
+   # You can use authentication API to create new users and grant permissions
    druid.auth.authenticator.MyBasicMetadataAuthenticator.credentialsValidator.type=metadata
 
-   # If true and the request credential doesn't exists in this credentials store, the request will proceed to next Authenticator in the chain.
+   # If true and the request credential doesn't exist in this credentials store,
```
[the remark this answers is not in the document]
Comment (abhishekrb19) [reply docs/operations/security-overview.md:155]: Yep, sounds good!

**Review by vtlim — APPROVED** [review]
LGTM

**Pull request merged**